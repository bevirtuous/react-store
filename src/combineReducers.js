/**
 * Combines an object of many reducers to one new higher order reducer.
 * @param {Object} reducers An object of multiple reducers.
 * @returns {Function} A new higher order reducer that unifies all the others.
 */
export function combineReducers(reducers = {}) {
  return (state = {}, action) => (
    Object.keys(reducers).reduce((nextState, key) => {
      // eslint-disable-next-line no-param-reassign
      nextState[key] = reducers[key](state[key], action);
      return nextState;
    }, {})
  );
}
