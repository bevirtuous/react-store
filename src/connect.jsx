import React from 'react';
import StoreContext from './context';

export default (mapStateToProps, mapDispatchToProps) => (
  Component => props => (
    <StoreContext.Consumer>
      {({ dispatch, getState }) => (
        <Component
          {...props}
          {...(mapStateToProps(getState(), props) || {})}
          {...(mapDispatchToProps(dispatch, props) || {})}
        />
      )}
    </StoreContext.Consumer>
  )
);
