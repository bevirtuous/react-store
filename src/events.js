import EventEmitter from 'events';

export const STORE_DISPATCH = 'STORE_DISPATCH';

export const emitter = new EventEmitter();
