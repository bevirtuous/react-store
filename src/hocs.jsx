import React from 'react';
import { StoreContext } from './context';

/**
 * Injects the store's dispatch function into the component props.
 * @param {Function} Component The original component.
 * @returns {Function}
 */
export function withDispatch(Component) {
  return function WithDispatch(props) {
    return (
      <StoreContext.Consumer>
        {({ dispatch }) => <Component {...props} dispatch={dispatch} />}
      </StoreContext.Consumer>
    );
  };
}

/**
 * Injects the store properties into the component props.
 * @param {Function} Component The original component.
 * @returns {Function}
 */
export function withStore(Component) {
  return function WithStore(props) {
    return (
      <StoreContext.Consumer>
        {store => <Component {...props} {...store} />}
      </StoreContext.Consumer>
    );
  };
}
