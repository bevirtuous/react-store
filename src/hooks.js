import { useContext } from 'react';
import StoreContext from './context';

/**
 * Returns the global application store's **dispatch** function.
 * @returns {Function} The global dispatch function.
 */
export const useDispatch = () => {
  const { dispatch } = useContext(StoreContext);

  return dispatch;
};

/**
 * Takes a **selector** and returns its output value.
 * @param {Function} selector The selector to exectue.
 * @param {Object} [props] Some component props to pass to the selector.
 * @returns {*} The output value of the selector.
 */
export const useSelector = (selector, props = undefined) => {
  const { getState } = useContext(StoreContext);

  return selector(getState(), props);
};

/**
 * Returns an instance of the store.
 * @returns {Object} The application store.
 */
export const useStore = () => {
  const store = useContext(StoreContext);

  return store;
};
