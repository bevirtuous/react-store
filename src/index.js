export { Store } from './Store';
export { useDispatch, useSelector, useStore } from './hooks';
export { withDispatch, withStore } from './hocs';
export { connect } from './connect';
export { combineReducers } from './combineReducers';
export { dispatch, select } from './helpers';
