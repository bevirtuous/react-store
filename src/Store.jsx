import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import StoreContext from './context';
import isPlainObject from './utils/isPlainObject';
import { emitter, STORE_DISPATCH } from './events';

export const storeData = {
  store: null,
};

/**
 * Creates a store that holds the state tree.
 * The only way to change the data in the store is to call `dispatch()` on it.
 *
 * There should only be a single store in your app. To specify how different
 * parts of the state tree respond to actions, you may combine several reducers
 * into a single reducer function by using `combineReducers`.
 */
export default class Store extends PureComponent {
  static propTypes = {
    /**
     * This usually represents your application contents. Everything that should be covered by
     * the store, should be a child or part of a child of the Store component.
     */
    children: PropTypes.node.isRequired,
    /**
     * A function that returns the next state tree, given the
     * current state tree and the action to handle.
     */
    reducer: PropTypes.func.isRequired,
    /**
     * The initial state. You may optionally specify it to hydrate the state from the server in
     * universal apps, or to restore a previously serialized user session. If you use
     * `combineReducers` to produce the root reducer function, this must be an object with
     * the same shape as `combineReducers` keys.
     */
    initialState: PropTypes.shape(),
    /**
     * An array of functions that extend the dispatch function. Note that each middleware
     * will be given the `dispatch` and `getState` functions as named arguments.
     */
    middlewares: PropTypes.arrayOf(PropTypes.func),
  }

  static defaultProps = {
    initialState: undefined,
    middlewares: [],
  }

  isDispatching = false;

  /**
   * @param {Object} props  The component props.
   */
  constructor(props) {
    super(props);

    this.state = props.reducer(props.initialState, {});
    storeData.store = this.applyMiddleware(props.middlewares);
    emitter.addListener(STORE_DISPATCH, this.dispatch);
  }

  componentDidMount() {
    // Initialize the store.
    this.dispatch({ type: '@@INIT' });
  }

  componentWillUnmount() {
    // Remove the dispatch listener.
    emitter.removeListener(STORE_DISPATCH, this.dispatch);
  }

  /**
   * Creates a store enhancer that applies middleware to the dispatch method
   * of the Redux store. This is handy for a variety of tasks, such as expressing
   * asynchronous actions in a concise manner, or logging every action payload.
   * @param {Function[]} middlewares A collection of middlewares.
   * @returns {Object} The enhanced store.
   */
  applyMiddleware = (middlewares) => {
    let dispatch = () => {
      throw new Error(
        'Dispatching while constructing your middleware is not allowed. '
        + 'Other middleware would not be applied to this dispatch.'
      );
    };

    const middlewareAPI = {
      getState: this.getState,
      dispatch: (...args) => dispatch(...args),
    };
    const chain = middlewares.map(middleware => middleware(middlewareAPI));
    dispatch = this.compose(...chain)(this.dispatch);

    return {
      getState: this.getState,
      dispatch,
    };
  }

  /**
   * Composes single-argument functions from right to left. The rightmost
   * function can take multiple arguments as it provides the signature for
   * the resulting composite function.
   *
   * @param {...Function} funcs The functions to compose.
   * @returns {Function} A function obtained by composing the argument functions
   * from right to left. For example, compose(f, g, h) is identical to doing
   * (...args) => f(g(h(...args))).
   */
  compose = (...funcs) => {
    if (funcs.length === 0) {
      return arg => arg;
    }

    if (funcs.length === 1) {
      return funcs[0];
    }

    return funcs.reduce((a, b) => (...args) => a(b(...args)));
  }

  /**
   * Dispatches an action. It is the only way to trigger a state change.
   * @param {Object} action A plain object representing "what changed". An action must have
   * a `type` property which may not be `undefined`. It is a good idea to use
   * string constants for action types.
   * @returns {Object} The new state.
   */
  dispatch = (action) => {
    if (!isPlainObject(action)) {
      throw new Error(
        'Actions must be plain objects. '
        + 'Use custom middleware for async actions.'
      );
    }

    if (typeof action.type === 'undefined') {
      throw new Error(
        'Actions may not have an undefined "type" property. '
        + 'Have you misspelled a constant?'
      );
    }

    if (this.isDispatching) {
      throw new Error('Reducers may not dispatch actions.');
    }

    const { reducer } = this.props;
    const { state } = this;

    this.isDispatching = true;

    const newState = reducer(state, action);

    this.setState(newState);
    this.isDispatching = false;

    return newState;
  }

  /**
   * Returns the state tree managed by the store.
   * @returns {Object} The current state tree of your application.
   */
  getState = () => this.state

  /**
   * @returns {JSX.Element}
   */
  render() {
    const { children } = this.props;

    return (
      <StoreContext.Provider value={storeData.store}>
        {children}
      </StoreContext.Provider>
    );
  }
}
