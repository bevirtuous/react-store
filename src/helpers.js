import { emitter, STORE_DISPATCH } from './events';
import { storeData } from './Store';

/**
 * Dispatches action **action** from anywhere in your codebase.
 * @param {Object} action A plain object representing "what changed". An action must have
 * a `type` property which may not be `undefined`. It is a good idea to use
 * string constants for action types.
 */
export const dispatch = (action) => {
  if (typeof action === 'undefined') {
    throw new Error('The dispatch helper can only be called with a valid action.');
  }
  emitter.emit(STORE_DISPATCH, action);
};

export const select = (selector, props) => {
  if (typeof selector === 'undefined') {
    throw new Error('The select helper can only be called with a valid selector');
  }

  return selector(storeData.store.getState(), props);
};
