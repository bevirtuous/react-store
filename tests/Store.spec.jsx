import React from 'react';
import { mount } from 'enzyme';
import Store from '../src/Store';
import combineReducers from '../src/combineReducers';
import { useSelector } from '../src/hooks';
import { dispatch } from '../src/helpers';

const counter = (state = 0, action) => {
  switch (action.type) {
    case 'INCREMENT':
      return state + 1;
    case 'DECREMENT':
      return state - 1;
    default:
      return state;
  }
};

const AppContent = () => {
  const count = useSelector(state => state.counter);

  return (
    <div>
      {count}
    </div>
  );
};

const App = () => (
  <Store reducer={combineReducers({ counter })}>
    <AppContent />
  </Store>
);

describe('<Store />', () => {
  it('should render', () => {
    const wrapper = mount(<App />);
    expect(wrapper.html()).toEqual('<div>0</div>');
    expect(wrapper).toMatchSnapshot();
  });
});
