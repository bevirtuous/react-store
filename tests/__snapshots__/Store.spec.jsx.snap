// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Store /> should render 1`] = `
<App>
  <Store
    middlewares={Array []}
    reducer={[Function]}
  >
    <AppContent>
      <div>
        0
      </div>
    </AppContent>
  </Store>
</App>
`;
